apps/react-playground/src/app/Widgets/MessagingPrototype/services/messaging-app.service.tsx

```diff
@@ -0,0 +1,163 @@
+import {
+  authorizeDerivedKey,
+  generateDefaultKey,
+  getEncryptedMessages,
+  login,
+  requestDerivedKey,
+} from './messaging.service';
+import {
+  getDerivedKeyResponse,
+  setDefaultKey,
+  setDerivedKeyResponse,
+} from './store';
+import { delay } from './utils';
+export const getConversationsMapAndPublicKeyToUsernameMapping = async (
+  deso: Deso,
+  derivedResponse: Partial<DerivedPrivateUserInfo>
+): Promise<{
+  conversationMap: DecryptedResponse;
+  publicKeyToUsername: { [key: string]: string };
+}> => {
+  if (!derivedResponse) {
+    alert('no derived response found');
+    return { conversationMap: {}, publicKeyToUsername: {} };
+  }
+  const messages = await getEncryptedMessages(deso);
+  const decryptedMessages = await deso.utils.decryptMessagesV3(
+    messages,
+    derivedResponse.messagingPrivateKey as string
+  );
+  const conversationMap: { [key: string]: any[] } = {};
+  const publicKeyToUsername: { [key: string]: string } = {};
+  const userKey = deso.identity.getUserKey();
+  Object.keys(decryptedMessages)?.forEach(async (key: string) => {
+    decryptedMessages[key].forEach((message) => {
+      const otherUsersKey =
+        userKey === message.RecipientPublicKeyBase58Check
+          ? message.SenderPublicKeyBase58Check
+          : message.RecipientPublicKeyBase58Check;
+      if (!conversationMap[otherUsersKey]) {
+        conversationMap[otherUsersKey] = [];
+      }
+      publicKeyToUsername[key] =
+        messages.PublicKeyToProfileEntry[key].Username ?? null;
+      conversationMap[otherUsersKey].push({ ...message });
+    });
+  });
+  return { conversationMap, publicKeyToUsername };
+};
+
+export const setupMessaging = async (
+  deso: Deso,
+  setDerivedResponse: (
+    derivedResponse: Partial<DerivedPrivateUserInfo>
+  ) => void,
+  setHasSetupAccount: (hasSetup: boolean) => void
+): Promise<false | Partial<DerivedPrivateUserInfo>> => {
+  let key = deso.identity.getUserKey();
+  if (!key) {
+    // are they already logged in? if not prompt them
+    const res = await login(deso);
+    key = res.key;
+    if (!key) {
+      alert('failed to login');
+      return false;
+    }
+  }
+  const userResponse = await deso.user.getUsersStateless({
+    PublicKeysBase58Check: [key],
+    SkipForLeaderboard: true,
+    IncludeBalance: true,
+  });
+  const user = userResponse?.UserList?.[0];
+  if (user && user.BalanceNanos === 0) {
+    // does the user have a balance? If not let them know since they will not be able to proceed
+    const openFreeDeso = window.confirm(
+      `no deso funds found for ${key}. click okay to add some through phone verification. Otherwise you can send deso from another account`
+    );
+    if (openFreeDeso) {
+      await deso.identity.phoneVerification('4', undefined, {
+        publicKey: deso.identity.getUserKey() as string,
+      });
+    } else {
+      return false;
+    }
+  } else if (!user) {
+    alert('unable to find user');
+    return false;
+  }
```

Comment:
```suggestion
  if (!user) {
    alert('unable to find user');
    return false;
  }
  if (user.BalanceNanos === 0) {
    // does the user have a balance? If not let them know since they will not be able to proceed
    const openFreeDeso = window.confirm(
      `no deso funds found for ${key}. click okay to add some through phone verification. Otherwise you can send deso from another account`
    );
    if (!openFreeDeso) {
      return false;
    }
    await deso.identity.phoneVerification('4', undefined, {
      publicKey: key,
    });
  }
```
Early returns keep cognitive complexity down